src/bin/pg_dump/dumpall_babel_utils.c

```diff
@@ -57,145 +57,118 @@ isBabelfishDatabase(PGconn *conn)
-					"ORDER BY 2", role_catalog, role_catalog);
-		return;
+		/*
+		 * Get escaped bbf_db_name to handle special characters in it.
+		 * 2*strlen+1 bytes are required for PQescapeString according to the documentation.
+		 */
+		char *escaped_bbf_db_name = pg_malloc(2 * strlen(bbf_db_name) + 1);
+
+		PQescapeString(escaped_bbf_db_name, bbf_db_name, strlen(bbf_db_name));
+		appendPQExpBuffer(buf, "WHERE database_name = '%s' ", escaped_bbf_db_name);
+		pfree(escaped_bbf_db_name);
 	}
+	appendPQExpBuffer(buf, "), "
+					  "bbf_roles AS (SELECT rc.* FROM %s rc INNER JOIN bbf_catalog bcat "
+					  "ON rc.rolname = bcat.rolname) ", role_catalog);
 
-	/* Modify role query for logical database dump. */
-	is_builtin_db = (pg_strcasecmp(bbf_db_name, "master") == 0 ||
-			pg_strcasecmp(bbf_db_name, "tempdb") == 0 ||
-			pg_strcasecmp(bbf_db_name, "msdb") == 0)
-			? true : false;
-	/*
-	 * Get escaped bbf_db_name to handle special characters in it.
-	 * 2*strlen+1 bytes are required for PQescapeString according to the documentation.
-	 */
-	escaped_bbf_db_name = pg_malloc(2 * strlen(bbf_db_name) + 1);
-	PQescapeString(escaped_bbf_db_name, bbf_db_name, strlen(bbf_db_name));
-
-	resetPQExpBuffer(buf);
 	if (drop_query)
 	{
-		printfPQExpBuffer(buf,
-						  "SELECT rolname "
-						  "FROM %s "
-						  "WHERE rolname !~ '^pg_' "
-						  "AND rolname IN ('dbo', 'db_owner', '%s_dbo', '%s_db_owner', '%s_guest') "
-						  "ORDER BY 1 ",
-						  role_catalog, escaped_bbf_db_name,
-						  escaped_bbf_db_name, escaped_bbf_db_name);
-
-		/* builtin db users will already be present in the target server so no need to dump them */
-		if (is_builtin_db)
-			appendPQExpBufferStr(buf, "LIMIT 0");
+		appendPQExpBufferStr(buf,
+							 "SELECT rolname "
+							 "FROM bbf_roles "
+							 "WHERE rolname !~ '^pg_' "
+							 "AND rolname NOT IN ('sysadmin', 'dbo', 'db_owner', "
+							 "'master_dbo', 'master_db_owner', 'master_guest', "
+							 "'msdb_dbo', 'msdb_db_owner', 'msdb_guest', "
+							 "'tempdb_dbo', 'tempdb_db_owner', 'tempdb_guest') "
+							 "ORDER BY 1 ");
 	}
 	else
 	{
-		printfPQExpBuffer(buf,
+		appendPQExpBuffer(buf,
 						  "SELECT oid, rolname, rolsuper, rolinherit, "
 						  "rolcreaterole, rolcreatedb, "
 						  "rolcanlogin, rolconnlimit, rolpassword, "
 						  "rolvaliduntil, rolreplication, rolbypassrls, "
 						  "pg_catalog.shobj_description(oid, '%s') as rolcomment, "
 						  "rolname = current_user AS is_current_user "
-						  "FROM %s "
+						  "FROM bbf_roles "
 						  "WHERE rolname !~ '^pg_' "
-						  "AND rolname IN ('dbo', 'db_owner', '%s_dbo', '%s_db_owner', '%s_guest') "
-						  "ORDER BY 2 ", role_catalog, role_catalog, escaped_bbf_db_name,
-						  escaped_bbf_db_name, escaped_bbf_db_name);
-
-		/* builtin db users will already be present in the target server so no need to dump them */
-		if (is_builtin_db)
-			appendPQExpBufferStr(buf, "LIMIT 0");
+						  "AND rolname NOT IN ('sysadmin', 'dbo', 'db_owner', "
```

Comment:
Same comment as last instance for use of 'dbo' and 'db_owner'.

Reply:
Same reason as in the reply above, removed both `dbo` and `db_owner` from here.